### ISSUE TITLE
What is the ideal procedure to merge scenes?

### ISSUE THREAD

**Issue by jpwhitney** [openMVG/openMVG#612]
I have two scenes that I am processing independently due to the large number of photos.  Eventually I'd like to merge these scenes.  Where is the best point in the pipeline to merge these two scenes?

The end result is to obtain a DEM from the 3D point cloud if that helps.

**Comment by pmoulon** [comment]
Hi @jpwhitney,
Does the scene have some images in common (i.e some image registered in the two reconstruction)?

Does your images have some GPS image data?

**Comment by jpwhitney** [comment]
At the moment the scenes do not have any images in common.  They were taken with two different grid sampling flight paths, but I believe there may be some overlap in the regions photographed.

Yes, the images have some GPS data, which brings up another question, will I still need to use GCP to georegister the resulting pointcloud?

@pmoulon btw thank you for your work

**Comment by pmoulon** [comment]
If the scenes does not have images in common we cannot compute a similarity between the reconstruction **but** we can do the following to align your datasets:
 - take the whole dataset (the two flight path and handle it as a single scene) (recomputing matches & SfM -> with the global SfM pipeline it will not be so long)
**or**
 - You can register each scene with the GCP
**or**
 - You can register each scene thanks to their GPS coordinates by using this binary (see the branch develop_geodesy)
https://github.com/openMVG/openMVG/issues/547#issuecomment-221589755

Tell me which strategy you want to try?

**Comment by jpwhitney** [comment]
The geodesy method seems the most attractive.  I'm trying to minimize user input in order to build a complete chain of commands from images to DEM that can then be loaded into GIS.

Which brings up another side question if DEM export is part of the openMVG toolkit or if not can you recommend another tool to take the openMVG scene to a DEM?

It seems that perhaps the Densify program of openMVS may be a useful step, but not sure if necessary.  Advice on the usefulness of densifying the point cloud before conversion to DEM would be greatly appreciated.

Some info about the data, the full set of photos is around 600 images, 4GB, global pipeline made point cloud with 300,000 points.

**Comment by jpwhitney** [comment]
I found the geodesy branch and will try to use the binary mentioned, however during the global script on the full image set RAM usage is 8 GB while swap usage is 3 GB and the operation grinds to a halt.
Perhaps this is a naive question but would running sequential before global change this?  I am unclear on the distinction between the functional usage of sequential vs. global and so far have not found a description of the difference.

**Comment by pmoulon** [comment]
Hi @jpwhitney 
Using Sequential or Global will not change the RAM problem.
I think the process use too much ram at the ComputeMatches step. Am I right?
It's seems that using all the image at the same time make having too much descriptor to store in memory in order to compute the image matches (necessary for incremental & global SfM)

In a WIP branch, I added a feature cache system in order to handle larger scene ( branch develop_regions_provider_cache)
https://github.com/openMVG/openMVG/blob/develop_regions_provider_cache/src/software/SfM/main_ComputeMatches.cpp#L90
You will see that there is an option -c X
X is the number of descriptor file that you allow to have in memory at the same time.
It make the process a bit slower, but you can handle larger scenes ;-)

FYI:
info about Incremental SfM: http://openmvg.readthedocs.io/en/latest/software/SfM/IncrementalSfM/
info about Global SfM: http://openmvg.readthedocs.io/en/latest/software/SfM/GlobalSfM/

Happy to see that you tackle some limits and that we have already new opportunity to handle your dataset!

**Comment by jpwhitney** [comment]
Thank you for your comments.

I will checkout the WIP and geodesy branches and let you know how it goes. ;)

**Comment by jpwhitney** [comment]
Hello @pmoulon,

I checked out the WIP branch and compiled.  I edited the SfM_GlobalPipeline.py file to include the '-c X' option for the compute matches step with X = 200.(see below for change)

`print ("3. Compute matches")
pMatches = subprocess.Popen( [os.path.join(OPENMVG_SFM_BIN, "openMVG_main_ComputeMatches"), **'-c 200'**  ,"-i", matches_dir+"/sfm_data.json", "-o", matches_dir, "-g", "e"] )
pMatches.wait()`

Everything seems to be going fine until I get some errors:

`3. Compute matches
 You called : 
/home/user/Applications/github/openMVG_WIP/Linux-x86_64-RELEASE/openMVG_main_ComputeMatches
--input_file /media/user/Seagate_4TB_USB/Rocky Shore Topographic Heterogeneity_DATA/PHOTOS_Drone/Processed_global/matches/sfm_data.json
--out_dir /media/user/Seagate_4TB_USB/Rocky Shore Topographic Heterogeneity_DATA/PHOTOS_Drone/Processed_global/matches
Optional parameters:
--force 0
--ratio 0.8
--geometric_model e
--video_mode_matching -1
--pair_list 
--nearest_matching_method AUTO
--guided_matching 0
--cache_size 200
Initialization of the Regions_Provider_Cache. #Elements in the cache: 200

 - PUTATIVE MATCHES - 
Use: exhaustive pairwise matching
Using FAST_CASCADE_HASHING_L2 matcher
Using the OPENMP thread interface

0%   10   20   30   40   50   60   70   80   90   100%
|----|----|----|----|----|----|----|----|----|----|
***************************************************
Task (Regions Matching) done in (s): 93822.8
Error: neato: can't open /media/user/Seagate_4TB_USB/Rocky
Error: neato: can't open Shore
Error: neato: can't open Topographic
Error: neato: can't open Heterogeneity_DATA/PHOTOS_Drone/Processed_global/matches/putative_matches
`

What happened?  Did I cause the error by editing the SfM_GlobalPipeline.py file the way that I did?  Also is this a recoverable error that can be fixed without all the processing time to run compute matches again?

The pipeline is continuing with Geometric filtering at present, will this be usable?

Sorry for so many questions, but they're all basically the same... :-)

I look forward to some clarification.

**Comment by jpwhitney** [comment]
Seems that perhaps the folder name having spaces was an issue...  I am running some tests now.

**Comment by pmoulon** [comment]
Since the OpenMVG tool log "Initialization of the Regions_Provider_Cache. #Elements in the cache: 200" I think the OpenMVG data were computed correctly and the cache sized was used.

The error message is only related to neato (an external tool I use to log visual graph on disk), and you are right this external does not like path with space. Normally OpenMVG must not stop if neato returned an error.

Once you have performed the matching does the process continue to process data or not?

**Comment by jpwhitney** [comment]
It continues to process with geometric filtering.

**Comment by pmoulon** [comment]
Ok, so only visual graph will not be exported.. Let's continue the experiment

**Comment by jpwhitney** [comment]
Ok, so I cancelled before I received your reply... Perhaps a hasty decision, but it seems I can continue with "#4. Do Global reconstruction", as the files created from "3. Compute matches" are still present.

**Comment by pmoulon** [comment]
The chain by default uses the previously computed file if they exists.
If you want to start from 4 go in the python script and comment the steps 1,2,3. Use the # character

**Comment by jpwhitney** [comment]
In case you were curious so far with my test of just 20 photos I have some good results.  
![image](https://cloud.githubusercontent.com/assets/17804207/17926053/52c00f7e-69a4-11e6-9d44-53e873827c4c.png)
![image](https://cloud.githubusercontent.com/assets/17804207/17926001/13e4f526-69a4-11e6-8032-6969b732d758.png)

**Comment by pmoulon** [comment]
Very nice results. Cannot wait to see the whole scene!
Thanks you so much for the sharing. I really appreciate to have an overview of the community reconstructions!

**Comment by jpwhitney** [comment]
Of course!  Also thanks for taking the time to assist me in this process.  I needed to keep #3 uncommented to continue, it is continuing through the pipeline now and I am eager for more results!  When I have some I will be sure to share them.  For now I will mark this question as closed.  I will reopen to share results and what I have discovered.

**Comment by jpwhitney** [comment]
![image](https://cloud.githubusercontent.com/assets/17804207/17994376/c43d0260-6b0c-11e6-8dfa-c280275aeafd.png)
![image](https://cloud.githubusercontent.com/assets/17804207/17994385/df18169c-6b0c-11e6-9ce3-5d97f5c63b27.png)

These are from the robust.ply output from the global pipeline python script.  As far as the procedure discussed above goes, the WIP branch's cache function works as expected and allowed the entire scene to be processed at one go, using about 5GB of RAM with cache size 200 for photos that average 5 MB, still waiting for an openMVS analog to the -c option... (densify fails, likely due to RAM issues), and the geodesy tool for georeferencing worked as expected as well, and for the test mesh, the registration gets carried through the openMVS processing as well.

Still researching how to convert the pointcloud to raster (without GUI) to be able to use in R or as a DEM in a GIS program.  But openMVG was definitely a valuable step in the process...  Thx again @pmoulon.

**Comment by pmoulon** [comment]
If you want to handle this scene in OpenMVS we can think about cut the scene in many clusters that you can process separately, then as each cluster is in a common coordinates system you can open all of them.

Really happy to hear that the beta features you tested was able to handle smoothly your dataset!
Seems like they are ready to be integrated into the develop branch (and so soon as a feature in a future release).

Thank you for your feedback & screenshot sharing. I really appreciate.

BTW, scene merging is still a problem that we can think about to solve in a future release.

**Comment by jpwhitney** [comment]
As far as processing clusters what I'm envisioning when you say that is to start the openMVG pipeline over with however many image clusters seem reasonable.  Am I interpreting you correctly here?

Assuming I am, I'm thinking of starting with 2 clusters with 1 row of overlapping photos on their mutual boundary.

It seems that PVMS is geared more for this type of clustering, do you have an opinion on using PVMS to do what you suggest?

The flight lines cover about a kilometer of coastline and we are hoping for cm scale resolution in the final raster.  Now that I consider it, due to our lack of access to large hardware perhaps the way to choose the cluster size is to consider the how processable our final image will be and then work backwards to identify an appropriate cluster size.

Do you think one row of image overlap is sufficient? @pmoulon

**Comment by jpwhitney** [comment]
Alternatively, and perhaps more useful is to extend the functionality of the openMVG to openMVS program to split the scene into clusters that could be processed individually with openMVG and then reassembled later.

Is this theoretically possible, given the current data_SfM.bin data structures?

If so I'd be willing to help if pointed in the right direction.

**Comment by jpwhitney** [comment]
I meant to say processed individually with openMVS.

**Comment by pmoulon** [comment]
I think implementing a solution similar to the below paper would ne useful: An Integer Linear Programming model for View Selection on Overlapping Camera Clusters 3DV 2014, Tokyo, Japan.

**Comment by jpwhitney** [comment]
Seems like it's all there in the paper.  Where would be the step in the existing pipeline to include such a concept?

**Comment by pmoulon** [comment]
1. Read the paper
 http://www.vision.ee.ethz.ch/~rhayko/paper/3dv2014_mauro_joint_selection_clustering.pdf
2. See if others from the community can be interested in working on this with you.

3. If the paper is too complex or if you did not have the time to implement it, we can try a greedy approach to select some image clusters.

**Comment by srivathsanmurali** [comment]
Hi,
I have implemented the clustering algorithm that is implemented on the paper for a project at work. I would be open to open source the code after some clean up and testing. It works with the openmvg sfm data and returns clusters in the openmvs format.

**Comment by pmoulon** [comment]
@srivathsanmurali It would be very kind.
In which form do you think making the new tool (a third party project that link to OpenMVG) or a new software included in OpenMVG?

**Comment by srivathsanmurali** [comment]
So currently, I have it as an stand alone tool and I wrote an exporter from openMVG to my tool.

**Comment by srivathsanmurali** [comment]
@pmoulon I do not implement the view selection part of the paper. I figured the openmvs view selection process is good enough. Hence I have only implemented the view clustering part.

**Comment by pmoulon** [comment]
@srivathsanmurali I'm agree. Having only the view clustering part is sufficient for our actual needs.

**Comment by jpwhitney** [comment]
That would be great to have some clustering implementation.  However I have resolved the memory issue that lead to this question by allocating more virtual RAM.

I can envision scenes that are too large to make my approach practical.  Perhaps creating a new issue to discuss the clustering would be more appropriate.  I am closing this issue since, for the moment, my reason for starting it has been resolved.
